File: scala/app/controllers/InventoryController.scala
package controllers

import javax.inject._
import play.api.mvc._
import models.InventoryRepository
import play.api.data.Form
import play.api.libs.json.Json
import play.api.data.Forms._

import scala.concurrent.{ExecutionContext, Future}

/**
  */
@Singleton
class InventoryController @Inject()(inventoryRepository: InventoryRepository,cc: ControllerComponents)(implicit ec: ExecutionContext) extends AbstractController(cc) {

  val inventoryForm: Form[CreateInventoryForm] = Form {
    mapping(
      "productId" -> number,
      "inventoryCount"-> number,
    )(CreateInventoryForm.apply)(CreateInventoryForm.unapply)
  }

  def getAll = Action.async(
    implicit request => (
      inventoryRepository.list().map(
        inventory => Ok(Json.toJson(inventory))
      )
      )
  )

  def getById(id: Int) = Action.async { implicit request =>
    val options = for {
      maybeInventory <- inventoryRepository.findById(id)
    } yield (maybeInventory)

    options.map { case (opt) =>
      opt match {
        case Some(inventory) => Ok(Json.toJson(inventory))
        case None => NotFound
      }
    }
  }

  def create = Action.async { implicit request =>
    inventoryForm.bindFromRequest.fold(
      errorForm => {
        Future.successful(BadRequest("Failed to create inventory."))
      },
      inventory => {
        inventoryRepository.create(
          inventory.productId,
          inventory.inventoryCount,
        ).map { inventory =>
          Created(Json.toJson(inventory))
        }
      }
    )
  }

  def update(id: Int) =
    Action.async(parse.json){
      implicit request =>
        inventoryForm.bindFromRequest.fold(
          _ => {
            Future.successful(BadRequest("Failed to update inventory."))
          },
          inventory => {
            inventoryRepository.update(models.Inventory(
              id,
              inventory.productId,
              inventory.inventoryCount
            )).map({ _ =>
              Ok
            })
          }
        )
    }

  def delete(id: Int) = Action.async(
    inventoryRepository.delete(id).map(_ => Ok(""))
  )

}

case class CreateInventoryForm(productId: Int, inventoryCount: Int)

File: app/src/controllers/InventoryController.ts
import Axios from "axios";
import Inventory from "../models/Inventory";

export type PostInventoryBody = {
    productId: number,
    inventoryCount: number,
}

const getAllInventories = async () => {
    try {
        const inventories: Inventory[] = (await Axios.get('/inventories')).data;
        return inventories;
    } catch (error) {
        return [];
    }
}

const getInventory = async (id: number | string) => {
    try {
        const single: Inventory = (await Axios.get(`/inventories/${id}`)).data;
        return single;
    } catch (error) {
        return null;
    }
}

const postInventory = async (body: PostInventoryBody) => {
    try {
        const resp = await Axios.post(
            '/inventories',
            body,
            {
                headers: {
                    'Content-Type': 'application/json',
                }
            }
        );
        if (resp.status !== 201)
            return false;
        return resp.data;
    } catch (error) {
        return false;
    }
}

const deleteInventory = async (id: number | string) => {
    try {
        const resp = await Axios.delete(`/inventories/${id}`)
        if (resp.status !== 200)
            return false;
        return true;
    } catch (error) {
        return false;
    }
}

export { getAllInventories, getInventory, postInventory, deleteInventory, };
